Select Case - C1145 case-expr shall be of type character, integer, or…

kiranktp opened the pull request:
C1145 : case-expr shall be of type character, integer, or logical

tskeith commented on lib/semantics/check-select-stmt.cc:
```diff
@@ -0,0 +1,28 @@
+  const auto &parsedExpr{std::get<parser::Scalar<parser::Expr>>(selectcaseStmt.t).thing};
+  if (const auto *expr{GetExpr(parsedExpr)}) {
+    if (auto type{expr->GetType()}) {
+      if (type->category() == TypeCategory::Real) {
```
What if the expression is complex or a derived type? Also, what if it is type-less (i.e. a BOZ literal)?

kiranktp replied:
Thanks. Check must be for character, integer and logical type instead of other remaining types. I will incorporate this change.

tskeith commented on lib/semantics/check-select-stmt.cc:
```diff
@@ -0,0 +1,28 @@
+    if (auto type{expr->GetType()}) {
+      if (type->category() == TypeCategory::Real) {
+        context_.Say(parsedExpr.source, // C1145
+            "Select case expression shall be of type character, integer, or logical"_err_en_US);
```
Other messages use "must be" rather than "shall be" in errors like this.

kiranktp replied:
I picked up the error message as is from standard's document. I will make it "must be".

tskeith commented:
Please format your code with clang-format 7.

sscalpone commented:
Hi @kiranktp.  Please add tests not only for more types, but for other conditions too, such as ranges, overlapping ranges, missing case values, and duplicate case values.

kiranktp commented:
> Hi @kiranktp.  Please add tests not only for more types, but for other conditions too, such as ranges, overlapping ranges, missing case values, and duplicate case values.

Hi, Those will be C1146, C1147, C1148, C1149 Constraints.  I will add them subsequently. I wanted this new checker to be in place first. Post that, others (sameeran/Anshu) from AMD will push in the changes for select Rank and select Type construct.

sscalpone commented:
@kiranktp  Would like us to merge this change now to ease that sharing with Sameeran and Anshu?

Sameeranjoshi commented:
Hi , using clang-format-7 from ubuntu packages, like below
clang-format-7  -style=llvm -dump-config > ~/.clang-format
clang-format-7  -style=llvm file.cc

1] I see many differences in the formatted code from the rest of the coding style followed in other files, is there anything more to change in .clang-format file?

Thanks

klausler commented:
> Hi , using clang-format-7 from ubuntu packages, like below
> clang-format-7 -style=llvm -dump-config > ~/.clang-format
> clang-format-7 -style=llvm file.cc
> 
> 1] I see many differences in the formatted code from the rest of the coding style followed in other files, is there anything more to change in .clang-format file?
> 
> Thanks

There is a `.clang-format` file in the topmost directory of the repository.

kiranktp commented:
> @kiranktp  Would like us to merge this change now to ease that sharing with Sameeran and Anshu?

I have inc

kiranktp commented:
> @kiranktp  Would like us to merge this change now to ease that sharing with Sameeran and Anshu?

Hi Steve, I have Incorporated review comments and added support for C1146, C1147, C1148, Request Tim and Klausler to review.
